docs/architecture/adr-073-libp2p.md

```diff
@@ -0,0 +1,189 @@
+# ADR 073: Adopt LibP2P
+
+## Changelog
+
+- 2021-11-02: Initial Draft (@tychoish)
+
+## Status
+
+Proposed.
+
+## Context
+
+As part of the 0.35 development cycle, the Tendermint team completed
+first phase of the work described in ADRs 61 and 62, which included a
+large scale refactoring of the reactors and the p2p message
+routing. This replaced the switch and many of the other legacy
+components, without breaking protocol or network-level
+interoperability and left the legacy connection/socket handling code.
+
+Following the release, the team has reexamined the state of the code
+and the design, as well as Tendermints requirements for. The notes
+from this process are available in the [P2P Roadmap
+RFC](../rfc/rfc-000-p2p.rst).
+
+While this ADR supersedes the decisions made in ADRs 60 and 61, it is
+also enabled by the completed portions of this work. Previously, the
+boundaries of peer management, message handling, and the higher level
+business logic (e.g. "the reactors") were intermingled, and core
+elements of the p2p system were responsible for the orchestration of
+the higher level business logic. The completed aspects of the refactor
+made it more obvious that the design implications of the legacy
+components had outsized influence on the entire implementation, that
+it would be difficult to iterate within the current abstractions, and
+it would not be viable to maintain interoperability with legacy
+systems while also achieving any of our broader objectives.
```

Comment:
Suggestions to break up the long sentence a bit.

```suggestion
higher-level business logic. Refactoring the legacy components
made it more obvious that this entanglement of responsibilities
had outsized influence on the entire implementation, making
it difficult to iterate within the current abstractions.
It would not be viable to maintain interoperability with legacy
systems while also achieving many of our broader objectives.
```